File: languages/typescript/src/smells/rules/no-unreachable-code.ts
import { defineRule } from "@oxlint/plugins";
import type { ESTree } from "@oxlint/plugins";

const TERMINATORS: Readonly<Record<string, string>> = {
	ReturnStatement: "return",
	ThrowStatement: "throw",
	BreakStatement: "break",
	ContinueStatement: "continue",
};

function reportUnreachable(
	body: readonly ESTree.Statement[],
	report: (node: ESTree.Statement, keyword: string) => void,
): void {
	for (let index = 0; index < body.length - 1; index += 1) {
		const statement = body[index];
		const keyword = statement === undefined ? undefined : TERMINATORS[statement.type];
		const following = body[index + 1];
		if (keyword === undefined || following === undefined) continue;
		report(following, keyword);
		return;
	}
}

/** Ban statements that can never run. */
export const noUnreachableCodeRule = defineRule({
	meta: {
		type: "problem",
		docs: {
			description:
				"Disallow statements after `return`, `throw`, `break`, or `continue` in the same block.",
		},
		messages: {
			unreachableCode:
				"Nothing after the `{{keyword}}` above can run, so this code is dead. Delete it, or move the `{{keyword}}` if the order is wrong.",
		},
	},
	createOnce(context) {
		const report = (node: ESTree.Statement, keyword: string) => {
			context.report({ node, messageId: "unreachableCode", data: { keyword } });
		};

		return {
			BlockStatement(node) {
				reportUnreachable(node.body, report);
			},
			SwitchCase(node) {
				reportUnreachable(node.consequent, report);
			},
			Program(node) {
				reportUnreachable(node.body as readonly ESTree.Statement[], report);
			},
		};
	},
});

File: languages/python/src/slop_guard/rules/smells/no_unreachable_code.py
"""Reject statements that can never run."""

from __future__ import annotations

import ast

from ...rule import Rule

TERMINATORS = (ast.Return, ast.Raise, ast.Break, ast.Continue)


class NoUnreachableCode(Rule):
    name = "no-unreachable-code"
    group = "smells"
    reference = "https://refactoring.guru/smells/dead-code"
    description = "Disallow statements after `return`, `raise`, `break`, or `continue` in the same block."

    def visit_Module(self, node: ast.Module) -> None:
        self._scan(node.body)
        self.generic_visit(node)

    def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
        self._scan(node.body)
        self.generic_visit(node)

    visit_AsyncFunctionDef = visit_FunctionDef

    def visit_For(self, node: ast.For) -> None:
        self._scan(node.body)
        self._scan(node.orelse)
        self.generic_visit(node)

    visit_AsyncFor = visit_For
    visit_While = visit_For
    visit_If = visit_For

    def visit_With(self, node: ast.With) -> None:
        self._scan(node.body)
        self.generic_visit(node)

    visit_AsyncWith = visit_With

    def visit_Try(self, node: ast.Try) -> None:
        self._scan(node.body)
        self._scan(node.orelse)
        self._scan(node.finalbody)
        for handler in node.handlers:
            self._scan(handler.body)
        self.generic_visit(node)

    def _scan(self, body: list[ast.stmt]) -> None:
        for index, statement in enumerate(body[:-1]):
            if isinstance(statement, TERMINATORS):
                following = body[index + 1]
                keyword = type(statement).__name__.lower()
                self.report(
                    following,
                    f"Nothing after the `{keyword}` above can run, so this code is dead. Delete "
                    f"it, or move the `{keyword}` if the order is wrong.",
                )
                return
